stub_uploader/upload.py

```diff
@@ -25,24 +25,43 @@ def upload(
     print("Building and uploading stubs for:", ", ".join(distributions))
     print()
     for distribution in to_upload:
-        print(f"Building stubs for {distribution}... ", end="")
-        metadata = read_metadata(typeshed_dir, distribution)
-        recursive_verify(metadata, typeshed_dir)
+        upload_distribution(typeshed_dir, distribution, commit, dry_run=dry_run)
 
+
+def upload_distribution(
+    typeshed_dir: str, distribution: str, commit: str | None, *, dry_run: bool
+) -> None:
+    print(f"Building stubs for {distribution}... ", end="")
+    metadata = read_metadata(typeshed_dir, distribution)
+    recursive_verify(metadata, typeshed_dir)
+
+    try:
         version = determine_stub_version(metadata)
-        if commit:
-            update_changelog(
-                typeshed_dir, commit, distribution, version, dry_run=dry_run
-            )
-
-        temp_dir = build_wheel.main(typeshed_dir, distribution, version)
-        print(f"ok, version {version}")
-
-        print(f"Uploading stubs for {distribution}... ", end="")
-        if dry_run or not metadata.upload:
-            print(f"skipped")
-        else:
-            subprocess.run(["twine", "upload", os.path.join(temp_dir, "*")], check=True)
-            uploaded_packages.add(distribution)
-            print(f"ok")
-        print()
+    except AlreadyUploadedError as e:
+        version = str(e.version)
+        already_uploaded = True
+    else:
+        already_uploaded = False
+
+    if commit:
+        update_changelog(
+            typeshed_dir,
+            commit,
+            distribution,
+            version,
+            dry_run=dry_run or already_uploaded,
+        )
+
+    temp_dir = build_wheel.main(typeshed_dir, distribution, version)
+    print(f"ok, version {version}")
+
+    print(f"Uploading stubs for {distribution}... ", end="")
+    if dry_run or not metadata.upload:
+        print("skipped")
+    elif already_uploaded:
+        print("already uploaded, skipped")
```

Comment:
I'm a little worried about this skip. This could result in us dropping updates to specific stub packages / the state being confusing, because it won't prevent update of `data/last_typeshed_commit.sha1`

Reply:
Good point. It shouldn't be a problem in practice, since the upload only runs once per day, but I'll investigate a bit more tomorrow if we can make this safer. Alternatively we could switch back to the plan of just advancing to the next free date. I.e. use 20231216 if 20231214 and 20231215 were already uploaded, even if it's still the 14th.